CFE-544 : Apply user defined labels & tags to all GCP resources created by OpenShift (#1217)
openshift/enhancements: enhancements/api-review/gcp_user_defined_labels.md

---
title: apply-user-defined-labels-to-all-gcp-resources-created-by-openShift
authors:
  - "@bhb"
reviewers:

>   - @patrickdillon

Reviewer: The linter is failing because of the unquoted `@` here. The names need to be in double quotes.

Author: Thank you @dhellmann !!

approvers:
  - "@jerpeter1" ## approver for CFE
api-approvers:
  - "@JoelSpeed" ## approver for api component
creation-date: 2022-07-12
last-updated: 2022-07-12
tracking-link:
  - https://issues.redhat.com/browse/OCPPLAN-8155
  - https://issues.redhat.com/browse/CORS-2249
see-also:
  - "enhancements/api-review/custom-tags-aws.md"
replaces:
  - N/A
superseded-by:
  - N/A
---

# Apply user defined labels to all GCP resources created by OpenShift

## Summary

This enhancement describes the proposal to allow an administrator of Openshift to 
have the ability to apply user defined labels to those resources created by Openshift 
in GCP.

## Motivation

Motivations include but are not limited to:

- Allow admin, compliance, and security teams to keep track of assets and objects 
  created by OpenShift in GCP.

### User Stories

- As an openshift administrator, I want to have labels added to all resources created 
  in GCP by Openshift.

### Goals

- The administrator or service (in the case of Managed OpenShift) installing OpenShift 
  can configure allowed number of user-defined labels in the Openshift installer generated

>   install config, which is referred and applied by installer and the in-cluster operators

Reviewer 2: ```suggestion
  install config, which is referenced and applied by installer and the in-cluster operators
```

GCP resource.

### Workflow Description

- An Openshift administrator requests to add required labels to all GCP resources 
  created by Openshift by adding it in `platform.gcp.userLabels`
- Openshift installer validates the labels defined in `platform.gcp.userLabels` and 
  adds these labels to all resources created during installation and also updates 
  `.status.platformStatus.gcp.resourceLabels` of the `infrastructure.config.openshift.io`
- In cluster operators refers `.status.platformStatus.gcp.resourceLabels` of the 
  `infrastructure.config.openshift.io` to add labels to the resources created later.

#### Variation [optional]

### API Extensions
Enhancement requires below modifications to the mentioned CRDs
- Add `userLabels` field to `platform.gcp` of the `installconfigs.install.openshift.io`
```yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  name: installconfigs.install.openshift.io
spec:
  versions:
  - name: v1
    schema:
      openAPIV3Schema:
        properties:
          platform:
            properties:
              gcp:
                properties:
                  userLabels:
                    additionalProperties:

>                       type: string

Reviewer 2: same comment as azure: not sure about the type here. it should be the equivalent of a map in go

Author: Yes, it will be of type `map[string]string`

                      will add as labels to all resources that it creates. Resources
                      created by the cluster itself may not include these labels.
                  type: object
```

- Add `resourceLabels` field to `status.platformStatus.gcp` 
  of the `infrastructure.config.openshift.io`
```yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  name: infrastructures.config.openshift.io
spec:
  versions:
  - name: v1
    schema:
      openAPIV3Schema:
        properties:
          status:
            properties:
              platformStatus:
                properties:
                  gcp:
                    properties:
                      resourceLabels:
                        description: resourceLabels is a list of additional labels to apply to GCP resources created for the cluster. See https://cloud.google.com/compute/docs/labeling-resources for information on labeling GCP resources. GCP supports a maximum of 64 labels per resource. OpenShift reserves 14 labels for its internal use, and allows 50 labels for user configuration.
                        type: array

>                         maxItems: 50

Reviewer 2: how was this limit determined?

Author: I thought since customer asked for support up to 40 tags in AWS, keeping max for user configuration would be ideal.

Author: To be aligned with the AWS implementation, updated the number of tags supported to 32 for user configuration and 32 for OCP's internal use, shared equally between the two.

`google_storage_bucket, google_compute_instance, google_dns_managed_zone, google_compute_image`

And below list of terraform GCP APIs support labeling resources in beta version.
`google_compute_address, google_compute_forwarding_rule`

Below list of GCP Resources does not support labeling.
`VPC Network, VM Instance Group, Subnetwork, Target Pool, DNS Zone, Records, Router, NAT Rules, Backend Service, Health Check`

API update example:
A local variable should be defined, which merges the default label and the user
defined GCP labels, which should be referred in the GCP resource APIs.
``` terraform
locals {
  labels = merge(
    {
      "kubernetes.io_cluster.${var.cluster_id}" = "owned"
    },
    var.gcp_extra_labels,
  )
}

resource "google_storage_bucket" "ignition" {
  ...
  labels = local.labels
}
```

For resources supporting labels in beta version below changes are required
``` terraform
provider "googlebeta" {
  credentials = var.gcp_service_account
  project     = var.gcp_project_id
  region      = var.gcp_region
}

resource "google_compute_address" "cluster_ip" {
  provider = googlebeta
  ...
  labels = local.labels
}

> ```

Reviewer 2: we don't currently use the googlebeta provider. We would need to have a discussion about whether there are any downsides to using that provider and whether those resources merit any disadvantages

Author: Thank you!! A discussion will be helpful, before asking customers if those two resources need to have labels and if yes what are the issues of using googlebeta.

Reviewer 2: By discussion, I mean if you could just include in the enhancement some details about the googlebeta provider. It seems like what would need to happen is we would take existing resources that are being created with the normal provider and would switch those to using [the googlebeta provider](https://registry.terraform.io/providers/hashicorp/google/latest/docs/guides/provider_versions). The kinds of questions we would want to explore are:

1. Would switching existing resources to a different provider have any noticeable effects?
2. In particular, would customers need to enable additional APIs/services to use this provider.

Author: Updated `Risks and Mitigations` section with more details on google-beta provider.

---
title: gcp_user_defined_labels
authors:
  - "@bhb"
reviewers:
  - "@patrickdillon" ## reviewer for installer component
  - "@JoelSpeed" ## reviewer for api and machine-api-provider-gcp components
  - "@dmage" ## reviewer for cluster-image-registry-operator component
  - "@akhil-rane" ## reviewer for cloud-credential-operator component
approvers:
  - "@jerpeter1" ## approver for CFE
api-approvers:
  - "@JoelSpeed" ## approver for api component
creation-date: 2022-07-12
last-updated: 2022-07-12
tracking-link:
  - https://issues.redhat.com/browse/OCPPLAN-8155
  - https://issues.redhat.com/browse/CORS-2249
see-also:
  - "enhancements/api-review/custom-tags-aws.md"
replaces:
  - N/A
superseded-by:
  - N/A
---

# Apply user defined labels to all GCP resources created by OpenShift

## Summary

This enhancement describes the proposal to allow an administrator of OpenShift to 
have the ability to apply user defined labels to those resources created by OpenShift 
in GCP.

## Motivation

Motivations include but are not limited to:

- Allow admin, compliance, and security teams to keep track of assets and objects 
  created by OpenShift in GCP.

### User Stories

> - As an openshift administrator, I want to have labels added to all resources created 

Reviewer 3: This user story is missing the why, the typical format is `As a <persona>, I want to <do something>, so that <I can achieve some goal>`, but we don't have the goal part included here, please add the goal to explain what the administrator will be able to achieve with the new feature

Author: comment incorporated.

  on the the GCP resources during creation.
- Labels must be applied at creation time, in an atomic operation. It isn't acceptable 
  to create an object and to apply labels post cluster creation.

### Non-Goals

- Management(update/delete) of resource labels post creation of cluster is out of scope.

## Proposal

A label of the form `kubernetes.io/cluster/<cluster_name>:owned` will be added to every
resource created by Openshift to enable administrator to differentiate the resources
created for Openshift cluster. An administrator is not allowed to add or modify the label 
having the prefix `kubernetes.io` or `openshift.io` in the name. 

New `userLabels` field will be added to `platform.gcp` of install-config for the user 
to define the labels to be added to the resources created by installer and in-cluster operators.

If `platform.gcp.userLabels` of install-config has any label defined, same will be added 
to all the gcp resources created by Openshift except, when the label validation fails 
to adhere to any of the below conditions
1. A label name must have minimum of 1 character and can have maximum of 63 characters.

> 2. A label value can have maximum of 63 characters.

Reviewer 3: The string `kubernetes.io/cluster/` is 22 characters, which leave 41 characters for the cluster name, what happens if the cluster name exceeds 41 characters?

Author: Cluster name [generated](https://github.com/openshift/api/blob/28228c0f01ab20e1d49d8369157fb171ad53e102/config/v1/types_infrastructure.go#L52-L55) by appending a five characters length random string(+1 for `-` used as delimiter) to the user inputted name is restricted to 27 characters, which is within the limit and is not a problem I think.
/cc @patrickdillon Could you please confirm, if my understanding is correct?

Reviewer 4: The installer team already is currently handling this (for a different problem) on destroy here: https://github.com/barbacbd/installer/blob/master/pkg/destroy/gcp/disk.go#L23

                              minLength: 1
                              pattern: ^[0-9a-z_-]+$
```

### Implementation Details/Notes/Constraints [optional]
Add a new field `resourceLabels` to `.status.platformStatus.gcp` of the 
`infrastructure.config.openshift.io` type. Labels included in the `resourceLabels` field 
will be applied to new resources created for the cluster by the in-cluster operators.

The `resourceLabels` field in `status.platformStatus.gcp` will be populated by the 
installer using the entries from `platform.gcp.userLabels` field of `install-config`.

`status.platformStatus.gcp` field of `infrastructure.config.openshift.io` is immutable.

All operators that create GCP resources (Cluster Infrastructure ,Storage ,Node ,
NetworkEdge ,Internal Registry ,CCO) will apply these labels to all GCP resources they create.

Below list of terraform GCP APIs to create resources should be updated to add user
defined labels and as well the openshift default label in the installer component.
`google_storage_bucket, google_compute_instance, google_dns_managed_zone, google_compute_image`

And below list of terraform GCP APIs support labeling resources in beta version.

> `google_compute_address, google_compute_forwarding_rule`

Reviewer 5: Yes, the beta provider can be skipped. It'd be better to not support GCP beta version as it not GA.

,NetworkEdge, Internal Registry) will apply these labels during resource creation.

The userLabels field is intended to be set at install time and is considered immutable. 
Components that respect this field must only ever add labels that they retrieve from this 
field to cloud resources, they must never remove labels from the existing underlying cloud 
resource even if the labels are removed from this field(despite it being immutable).

If the userLabels field is changed post-install, there is no guarantee about how an 
in-cluster operator will respond to the change. Some operators may reconcile the 
change and update the labels on the GCP resource. Some operators may ignore the change. 
However, if labels are removed from userLabels, the label will not be removed from the 
GCP resource.

### Workflow Description

- An Openshift administrator requests to add required labels to all GCP resources 
  created by Openshift by adding it in `platform.gcp.userLabels`
- Openshift installer validates the labels defined in `platform.gcp.userLabels` and 
  adds these labels to all resources created during installation and also updates 

>   `.status.platformStatus.gcp.resourceLabels` of the `infrastructure.config.openshift.io`

Reviewer 3: This field is on status for AWS, however, the AWS project has now been updated and they are adding the field to the spec, so that users can update the tags after cluster install should they so desire.

I think if we add this to the status, we will shoot ourselves in the foot since we will then also need to add the the spec later. I'm wondering if we should add this directly to spec and document that updates aren't supported on all components, until we can find a way to make sure they are.

CC @TrilokGeer as SME on the AWS tagging updates

Reviewer 3: This comment can be resolved, will move forward with status

Author: Thank you for confirming.

by OpenShift in GCP.

## Motivation

Motivations include but are not limited to:

- Allow admin, compliance, and security teams to keep track of assets and objects
  created by OpenShift in GCP.

### User Stories

- As an openshift administrator, I want to have labels added to all resources created
  in GCP by Openshift, so that I can query resources by the labels or filter resources
  by label in cloud billing.
- As an openshift administrator, I want to have tags added to all resources created
  in GCP by Openshift, so that I can use it for creating IAM policies or to configure 
  organization policy constraints or to refine resource usage and cost data.

### Goals

- The administrator or service (in the case of Managed OpenShift) installing OpenShift
  can configure a list of up to 32 user-defined labels in the Openshift installer generated
  install config, which is referenced and applied by installer and the in-cluster operators
  on the the GCP resources during creation.
- Labels must be applied at creation time, in an atomic operation. It isn't acceptable
  to create an object and to apply labels post cluster creation.

> - The administrator or service (in the case of Managed OpenShift) installing OpenShift

Reviewer 4: Should we consider breaking out tags vs labels into two different enhancements? They aren't set to achieve the same goal so should they be grouped together?

Author: GCP provides tags and labels which is just tags having both functionalities in AWS, Azure. And since we were updating metadata of a resource(GCP tags aren't actually resource metadata but could be viewed like one) and getting reviewed too will be helpful, so thought to have single proposal for both. Should I break it into two proposals?

Reviewer 4: I am fine leaving them together if others are. They do go together although they are used for different purposes.

>   can configure a list of up to 50 user-defined tags in the Openshift installer generated

Reviewer 5: ```suggestion
  can configure up to 50 tags in the install config
```

>   install config, which is referenced and applied by installer and the in-cluster operators

Reviewer 5: ```suggestion
  The tags are added by installer and the in-cluster operators
```

>   on the the GCP resources during creation.

Reviewer 5: ```suggestion
  while creating GCP resources.
```

### Non-Goals

> - Management(update/delete) of resource labels or tags post creation of cluster is out of scope.

Reviewer 5: ```suggestion
- Support for update or delete operation on labels or tags after creation of GCP resource.
```

  Below are some of the risks in using the google-beta provider.
  1. google-beta is a preview version and doesn't always carry the SLAs or technical
  support obligations.
  2. google-beta provider always uses the beta API regardless of whether the
  request contains any beta features.
  3. google_compute_address and google_compute_forwarding_rule APIs provide just the
  `labels` functionality in the beta version, which is in beta since the initial
  [commit](https://github.com/hashicorp/terraform-provider-google-beta/commit/0fdc262183ed82e78824e42d05042a7b0aba7c8b) in 2018, though the usual beta
  phase for a feature is six months.

### Drawbacks
- User-defined labels cannot be updated on an GCP resource which is not managed by an
  operator. In this proposal, the changes proposed and developed will be part of
  openshift-* namespace. External operators are not in scope.
  User-defined labels can be updated on the following GCP resources.
  - Compute Engine Instance
  - Compute Address
  - Compute Image
  - Compute Forwarding Rule
  - Storage Bucket
  - DNS Zone

- GCP Labels can be used as queryable annotations for resources, but can't be used
  to set conditions on policies.

> - User-defined tags cannot be updated on an GCP resource which is not managed by an

Reviewer 4: ```suggestion
- User-defined tags cannot be updated on a GCP resource which is not managed by an
```

resource created by OpenShift to enable administrator to differentiate the resources
created for OpenShift cluster. An administrator is not allowed to add or modify the label
having the prefix `kubernetes-io` or `openshift-io` in the key. The same can be found
applied to other cloud platform resources which supports labels/tags for ex: AWS, Azure.

New `userLabels` field will be added to `platform.gcp` of install-config for the user
to define the labels to be added to the resources created by installer and in-cluster operators.

If `platform.gcp.userLabels` in the install-config has any labels defined, then these labels
will be added to all the GCP resources created by OpenShift, provided all the defined
labels meet all the below conditions or else the cluster creation will fail.
1. A label key and value must have minimum of 1 character and can have maximum of 63 characters.
2. A label key and value must contain only lowercase letters, numeric characters,
   underscores, and dashes.
3. A label key must start with a lowercase letter.
4. Each resource can have multiple labels, up to a maximum of 32.
   - GCP supports a maximum of 64 labels per resource. Restricting the number of

>     user defined labels to 32 and reserving 32 for Openshift's internal use.

Reviewer 3: This is useful context, thanks for including this

apply these labels during resource creation.

The userLabels field is intended to be set at install time and is considered immutable.
Components that respect this field must only ever add labels that they retrieve from this
field to cloud resources, they must never remove labels from the existing underlying cloud
resource even if the labels are removed from this field(despite it being immutable).

If the userLabels field is changed post-install, there is no guarantee about how an
in-cluster operator will respond to the change. Some operators may reconcile the
change and update the labels on the GCP resource. Some operators may ignore the change.
However, if labels are removed from userLabels, the label will not be removed from the
GCP resource.

***GCP Tags*** 

Unlike labels, Tags are not metadata of a resource but a resource itself. Tags are key-value
pairs. Tag keys, values, and bindings are all discrete resources. A tag key resource must be
created under an organization resource. A tag value created for each key is a list of values
that can be assigned to the key. A maximum of 1000 keys can be created under a given

> organization and there can be a total of 1000 values created for each key.

Reviewer 3: Are these absolute limits or can they be increased by a request to GCP? How will we ensure customers understand this since it will put a limit on the number of clusters that can be created? 
Do we know how many keys an openshift cluster will consume by default?

Author: Yeah, those are current limits for tags. I will check if those can be extended, but I didn't come across about the same in their docs.

Since these tags must already exist before cluster creation, openshift won't be adding any tags.

Reviewer 3: Can the tags be shared among clusters, or are they likely to be unique per cluster? If they are shared then this probably isn't an issue

Author: Yeah, it can be shared among clusters. Any tag added to a parent resource will be inherited by all the child resources. If a tag is added to a project, then all resources created for all the clusters under this project will inherit the tags. And each resource can be added with new tags and inherited tags can be modified.

Reviewer 3: Ok, in that case I'm not too worried about the limits, but we should make sure they are documented when the feature is documented

  openshift-* namespace. External operators are not in scope.
  User-defined labels can be updated on the following GCP resources.
  - Compute Engine Instance
  - Compute Address
  - Compute Image
  - Compute Forwarding Rule
  - Storage Bucket
  - DNS Zone

- GCP Labels can be used as queryable annotations for resources, but can't be used
  to set conditions on policies.

- User-defined tags cannot be updated on a GCP resource which is not managed by an
  operator. In this proposal, the changes proposed and developed will be part of
  openshift-* namespace. External operators are not in scope.
  User-defined tags can be updated on the following GCP resources.
  - Compute Engine Instances
  - Cloud Storage Buckets

- Administrator will have to manually perform below label pertaining actions
    1. removing the undesired labels/tags from the required resources.
    2. update label/tags values of the required resources.
    3. update labels/tags of the resources which are not managed by an operator.
    4. update labels/tags of the resources for which update logic is not supported
       by an operator.

## Design Details

> ### Open Questions

Reviewer 3: Do we need to make decisions on these open questions before merging the enhancement?

Author: Yeah I think it would be helpful. Atleast for TP if we can decide on what's best performance wise, we could update based on any user feedback for GA.

Author: `Open Questions` section updated with the solution considered for TechPreview.

      - To derive the tag value's `NamespacedName` using the aforementioned user provided parameters.
      - Use the derived tag value's `NamespacedName` to create tag binding to a resource using `CreateTagBinding` API.

  Another approach is for user to configure just the list of tag value's names of the form `tagValues/<TAG_VALUE_ID>`,
  and it can be used directly with both terraform and GO APIs, and no additional processing on tags would be required.
  But the tag values configured are in obscured form and it will be diffcult for mapping the tags configured and the
  tags present on resources.

  Would it be better to ask user to configure just the tag value name, and not Organization ID, tag key short name, though
  it reduces the readability?

### Test Plan
- Upgrade/downgrade testing
- Sufficient time for feedback
- Available by default
- Stress testing for scaling and label/tag update scenarios

### Graduation Criteria
#### Dev Preview -> Tech Preview
- Feature available for end-to-end usage.
- Complete end user documentation.
- UTs and e2e tests are present.
- Gather feedback from the users.

#### Tech Preview -> GA

> N/A. This feature is for Tech Preview, until decided for GA.

Reviewer 3: What featuregate will be used to gate this? Are we adding a new FeatureGate under the TechPreviewNoUpgrade FeatureSet? What testing will be in place before this is promoted to GA?

Author: We will add a TechPreviewNoUpgrade validation in installer [here](https://github.com/openshift/installer/blob/master/pkg/types/validation/installconfig.go#L1031) to allow configuration of tags and labels only when TP feature is set.

Full fledged testing will be carried out for TP implementation, any user suggestions/feedback incorporated will be tested for GA along with the regression tests. Please recommend if other tests are required or need to be looked at.

Reviewer 3: Features should be blocked by `FeatureGates` not `FeatureSets`, otherwise a use cannot turn on/customise their selection to individual features. Also, the controllers in the cluster need to be able to understand if the feature is enabled or not. We don't want them to run the tag syncing code if it's not enabled.

So you should be adding a [FeatureGate](https://github.com/openshift/api/blob/master/config/v1/feature_gates.go) and blocking the functionality on that rather than on `TechPreviewNoUpgrade`

Author: Since tags and labels can be configured only during cluster creation, through `FeatureSets`, I was thinking if `TechPreviewNoUpgrade` is not set, tags/labels configuration will not be allowed and cluster creation will fail with error.

And for operators since Infrastructure would not have tags & labels it will be skipped.
But a provision for configuring tags in machine provider operator like for AWS, Azure tags would have to be controlled through `FeatureGates` as it's not supported currently, though operator supports configuring labels.

Reviewer 3: By relying on the `FeatureSet`, you cannot allow users to customise their install. What if, for example, the user wanted to try this feature without other `TechPreviewNoUpgrade` feature gates? Normally they would do that using `CustomNoUpgrade` and selecting only the features they want, with this approach, that won't work

Author: Yeah, I understand, I will make changes to use `FeatureGates`

  to set conditions on policies.

- User-defined tags cannot be updated on a GCP resource which is not managed by an
  operator. In this proposal, the changes proposed and developed will be part of
  openshift-* namespace. External operators are not in scope.
  User-defined tags can be updated on the following GCP resources.
  - Compute Engine Instances
  - Cloud Storage Buckets

- Administrator will have to manually perform below label pertaining actions
    1. removing the undesired labels/tags from the required resources.
    2. update label/tags values of the required resources.
    3. update labels/tags of the resources which are not managed by an operator.
    4. update labels/tags of the resources for which update logic is not supported
       by an operator.

## Design Details

### Open Questions
- GCP supports 50 tags per resource and since OpenShift do not add any tags of its own,
  users are allowed to define upto 50 tags. But the terraform and GO APIs for binding the
  tags to the resources doesn't support batch operation and requires each tag to be added
  individually to each resource which will increase the cluster creation time. What would

>   be the optimal number of tags a user can define?

Reviewer 4: It looks like the google APIs limit the number of "requests" to 10 per second. If the tag additions cannot be batched we may need some timing operations to ensure that this quota isn't breached, or limit the number to 10.

Author: I was unaware of this, Thank you for sharing!
Yeah, I think limiting number of tags to 10 and timing the operations would be better. We did ask for a feedback from users, but no response yet. If more number of tags are required we could work on it while making this feature GA.

/cc @JoelSpeed @TrilokGeer

Reviewer 5: Agree with @JoelSpeed , RPS should not be a limitation for total tags. It'd be helpful to note this limitation in product documentation. There might be an added installation delay when there are large number resources being configured at installation. For example, worker nodes being configured for installation (approximately 5s per resource instance). Would it affect any timeout configured for installer?

Author: @JoelSpeed I am still looking for rate limiting code in k8s/openshift. I also found a sub-repo of golang which provides has the package `golang.org/x/time/rate` implementing rate limit functionality, using which I was able to limit the number of requests made in a second.

Author: `Open Questions` section updated with the solution considered for TechPreview.

Reviewer 3: Ok so we can rate limit on an individual component basis, but how does this work overall? If two operators are both rate limiting themselves, there is still a chance to hit the limit on the Google side right? I guess we can be best effort, but I'd be interested to know how the error comes back and if the operators are able to handle those well, any thoughts? Perhaps we don't need a rate limit ourselves if we can interpret the google errors sufficiently well

Author: Google errors does provide information on exceeding rate limit through http code 429 and the APIs support retry mechanism on particular http codes. I think it would be good to have both rate limiting mechanism of 8 requests/second and retry logic for http code 429, with initial backoff of 90seconds, since quota defined for Tag write operations is 600requests/minute and for maximum of 5minutes with retry period factor of 2. I did try out this after hitting the rate limit by adding 50 tags through 9 parallel routines and worked as expected, without rate limit exceeded errors.

Reviewer 3: Ok cool, as long as we have a plan, that sounds good, can you make sure to incorporate those details into the enhancement

Author: Yeah, I have captured it [here](https://github.com/openshift/enhancements/pull/1217/files#diff-0c4f04b9b4a7dd08afbc3b7fcf8c4a15b3b4b31b51faaf43ecb33c0e9de4cdc4R353-R357)

- GCP client libraries support two ways of providing tag details

>   1. NamespacedName: Tag value could be fed to APIs in the format

Reviewer 4: This will require us to add the organization ID to the install config. This seems better than more queries to the api to get the TAG ID for each tag.

Author: I agree going with the NamespacedName means, we would need to make
(configured_tag_count * 2[fetch tag key + tag value]) + (configured_tag_count * resources_to_tag) and if TagValue Names is used it would be just  (configured_tag_count * resources_to_tag). And configuring an additional `Organization_ID` parameter wouldn't be required.

/cc @TrilokGeer @JoelSpeed

Reviewer 5: I feel, it'd be better to maintain input uniformity in both install config and infra status when compared to other platforms unless there is significant delay or cost addition.

Author: To maintain uniformity with the solution provided for Aws, Azure tags, user will have to input Organization ID, tag key shortname and tag value shortname. The operators will use these parameters to derive the tags NamespacedName(<Organization_ID>/<tag_key_short_name>/<tag_value_short_name>)  and installer will use the tag value name fetched using the inputted params.

Author: `Open Questions` section updated with the solution considered for TechPreview.

Reviewer 2: This implementation is focusing on creating tags at the organization level, but typically openshift is more concerned with the project-level and it does seem possible to create tags at the project level. 

I would not expect the installer to have permissions to create tags at the organization level.

Author: OpenShift will not create tags, will add the tags already existing at organization level to the resources(VM, storage bucket) created for a cluster. OpenShift will also not add tags to the project in which the resources are created.

Openshift will require permissions for viewing the tags existing at Organization level and the SA created for resources will have the permissions for adding tags to the resources.

  1. For Terraform API([API doc](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/tags_tag_binding))
      - To fetch the tag key's `Name` using the `short_name`.
      - To fetch the tag value's `Name` using the `short_name` and the tag key's `Name`.
      - Use the tags value's `Name` to create tag binding to a resource using `google_tags_location_tag_binding` API.
  2. For GO API([API doc](https://github.com/googleapis/google-cloud-go/blob/resourcemanager/v1.8.2/resourcemanager/apiv3/tag_bindings_client.go#L174))
      - To derive the tag value's `NamespacedName` using the aforementioned user provided parameters.
      - Use the derived tag value's `NamespacedName` to create tag binding to a resource using `CreateTagBinding` API.

  Another approach is for user to configure just the list of tag value's names of the form `tagValues/<TAG_VALUE_ID>`,
  and it can be used directly with both terraform and GO APIs, and no additional processing on tags would be required.
  But the tag values configured are in obscured form and it will be diffcult for mapping the tags configured and the
  tags present on resources.

>   Would it be better to ask user to configure just the tag value name, and not Organization ID, tag key short name, though

Author: With the TagViewer permission at organization level, a user can make use of GCP cli to describe the tags like below , and user would require additional permissions to query in the console.
`gcloud resource-manager tags values describe tagValues/281483034419673`

created for OpenShift cluster. An administrator is not allowed to add or modify the label
having the prefix `kubernetes-io` or `openshift-io` in the key. The same can be found
applied to other cloud platform resources which supports labels/tags for ex: AWS, Azure.

New `userLabels` field will be added to `platform.gcp` of install-config for the user
to define the labels to be added to the resources created by installer and in-cluster operators.

If `platform.gcp.userLabels` in the install-config has any labels defined, then these labels
will be added to all the GCP resources created by OpenShift, provided all the defined
labels meet all the below conditions or else the cluster creation will fail.
1. A label key and value must have minimum of 1 character and can have maximum of 63 characters.
2. A label key and value must contain only lowercase letters, numeric characters,
   underscores, and dashes.
3. A label key must start with a lowercase letter.
4. Each resource can have multiple labels, up to a maximum of 32.
   - GCP supports a maximum of 64 labels per resource. Restricting the number of
    user defined labels to 32 and reserving 32 for Openshift's internal use.

> All in-cluster operators that create GCP resources (Cluster Infrastructure ,Storage) will

Author: OpenShift's `GCPMachineProviderSpec` has provision for defining labels [here](https://github.com/openshift/api/blob/master/machine/v1beta1/types_gcpprovider.go#L96) and is currently adding OCP default label to worker machines [here](https://github.com/openshift/machine-api-provider-gcp/blob/main/pkg/cloud/gcp/actuators/machine/reconciler.go#L224-L227), which would be extended to included user-defined labels.

And cluster-api is adding defined labels to Compute Instance [here](https://github.com/kubernetes-sigs/cluster-api-provider-gcp/blob/e65ddc123d01a9b6af84b85af650452035de5708/cloud/scope/machine.go#L341C3-L347)

> All in-cluster operators that create GCP resources (Cluster Infrastructure ,Storage) will

Reviewer 3: We need to make sure if we are adding functionality, that it matches in both Cluster API Provider GCP upstream, and the downstream

> apply these labels during resource creation.

Reviewer 6: As for storage, we already do label the volumes (PD: https://github.com/openshift/gcp-pd-csi-driver-operator/pull/16, Filestore: (https://github.com/openshift/gcp-filestore-csi-driver-operator/pull/38). This however requires support also from the CSI drivers, i.e. needs some upstream changes too. I assume it may be similar also for other resource types.

Author: Added as limitation in TP.

Reviewer 6: OK. I agree that the labels are fine and we know how to implement that change, so no problem in 4.14 with labels. Tags might be more complicated, so let's postpone that part to 4.15.

The userLabels field is intended to be set at install time and is considered immutable.
Components that respect this field must only ever add labels that they retrieve from this
field to cloud resources, they must never remove labels from the existing underlying cloud

> resource even if the labels are removed from this field(despite it being immutable).

Reviewer 6: The Filestore driver adds a `storage_gke_io_created-by: filestore_csi_storage_gke_io` label automatically. I'm not sure we would be able to change this behaviour since it's already in the driver and users may depend on it.

If the userLabels field is changed post-install, there is no guarantee about how an
in-cluster operator will respond to the change. Some operators may reconcile the
change and update the labels on the GCP resource. Some operators may ignore the change.
However, if labels are removed from userLabels, the label will not be removed from the
GCP resource.

> ***GCP Tags*** 

Reviewer 3: One thing that's unclear to me, and perhaps you can clarify, if I have a tag, and I want to apply that to say a VM, can I apply multiple values for a tag to the VM, or only one value per tag?

Author: We can tag a VM with only one value for a key, it's scalar.

Reviewer 3: Ok, please make sure to add a comment into the enhancement to explain that

Author: comment incorporated.

- OpenShift installer validates the labels defined in `platform.gcp.userLabels` and
  adds these labels to all resources created during installation and also updates
  `.status.platformStatus.gcp.resourceLabels` of the `infrastructure.config.openshift.io`
- In-cluster operators refers `.status.platformStatus.gcp.resourceLabels` of the
  `infrastructure.config.openshift.io` to add labels to the resources created later.

***GCP Tags***
- An OpenShift administrator requests to add required tags to all GCP resources
  created by OpenShift by adding it in `platform.gcp.userTags`
- OpenShift installer validates the tags defined in `platform.gcp.userTags` and
  adds these tags to all resources created during installation and also updates
  `.status.platformStatus.gcp.resourceTags` of the `infrastructure.config.openshift.io`
- In-cluster operators refers `.status.platformStatus.gcp.resourceTags` of the
  `infrastructure.config.openshift.io` to add tags to the resources created later.

#### Variation [optional]

### API Extensions
Enhancement requires below modifications to the mentioned CRDs
- Add `userLabels` and `userTags` fields to `platform.gcp` of the 

>   `installconfigs.install.openshift.io`

Reviewer 6: For persistent volumes this means the labels / tags could not be set by the driver (since this CRD is OpenShift specific), so something else would have to copy the labels from the CRD.

                  organizationID:
                    description: OrganizationID is the the organization identifier
                      that will be used for deriving the tag key's namespaced name.
                    type: string
                  userLabels:
                    additionalProperties:
                      type: string
                    description: UserLabels has additional keys and values that the installer
                      will add as labels to all resources that it creates on GCP. Resources
                      created by the cluster itself may not include these labels.
                    type: object
                  userTags:
                    additionalProperties:
                      type: string
                    description: UserTags has additional keys and values that the installer
                      will add as tags to all resources that it creates on GCP. Resources
                      created by the cluster itself may not include these tags. Tag key
                      and tag value should be the shortnames of the tag key and tag value
                      resource.
                    type: object
```

> - Add `resourceLabels` and `resourceTags` fields to `status.platformStatus.gcp`

Reviewer 3: Can you swap this schema for the Go definition, it's much easier to read and review than the schema

Author: comment incorporated.

                            rule: self.all(x, x in oldSelf) && oldSelf.all(x, x in self)
                      resourceTags:
                        description: resourceTags is a list of additional tags to apply to GCP resources created for the cluster. See https://cloud.google.com/resource-manager/docs/tags/tags-overview for information on tagging GCP resources. GCP supports a maximum of 50 tags per resource.
                        items:
                          description: GCPResourceTag is a tag to apply to GCP resources created for the cluster.
                          properties:
                            key:
                              description: key is the key part of the tag. A tag key can have a maximum of 63 characters and cannot be empty. Tag key must begin and end with an alphanumeric character, and must contain only uppercase, lowercase alphanumeric characters, and the following special characters `. _ -`.
                              maxLength: 63
                              minLength: 1
                              pattern: ^[a-zA-Z0-9]([0-9A-Za-z_.-]{0,61}[a-zA-Z0-9])?$
                              type: string

>                             value:

Reviewer 2: I believe, technically, GCP tags have a 1:many relationship between key:value, but this install config model has a 1:1 relationship. Could this potentially be problematic? Could users want to supply more than one value per key?

Author: We can create a maximum of 1000 keys and for each key 1000 values can be created. But OpenShift will not create tags, and only add tags to the resources it creates, I think tags with key and a value should suffice.

Author: When creating a tag key, multiple tag values can be defined for a key and user can choose one of these predefined values only for a key when attaching to a resource. User cannot assign any random key or value to a resource, but has to choose from one of the predefined keys and the predefined values for that key.

Below example lists the values defined for the key `ocp_tag_dev` created in Organization having ID `54643501348`
```
$ gcloud resource-manager tags values list --parent=54643501348/ocp_tag_dev
NAME                       SHORT_NAME  DESCRIPTION
tagValues/281478803580207  bar
tagValues/281483034419673  foo
tagValues/281484411361997  baz
```
Add a user to add a tag with key `ocp_tag_dev` then can choose one of `foo, bar, baz`as the value for the tag.
```
$ gcloud resource-manager tags bindings create --parent=//compute.googleapis.com/projects/openshift-gce-devel/zones/asia-south1-a/instances/7120657451740199772 --tag-value=54643501348/ocp_tag_dev/foo --location=asia-south1-a
```
And to add the same through OpenShift, will have to provide list of key:value pairs in the install-config.

Author: We can tag a resource with only one value for a key, which is documented [here](https://cloud.google.com/resource-manager/docs/tags/tags-creating-and-managing#attaching). We can tag a resource with the key:value pair `ocp_tag_dev:foo` or `ocp_tag_dev:bar` or `ocp_tag_dev:baz`. The install-config would look like below to tag resources with value `foo`
```
platform:
  gcp:
    organizationID: 54643501348
    userTags:
      ocp_tag_dev: foo
```

Reviewer 3: Ok, making more sense now thank you

I think the API should actually be as below to be consistent with kube standards right?
```
platform:
  gcp:
    organizationID: 54643501348
    userTags:
     - name: ocp_tag_dev
       value: foo
```

Author: Yeah, tags in Infrastructure will be an array of struct type with `key` and `value` as members, in install-config it will be type `map[string]string`.

Reviewer 3: Why not make them the same?

Author: I have observed for key:value parameters in install-config go maps are used and same is the type for AWS, Azure tags as well and decided to keep the same to be uniform across solutions. Received a [comment](https://github.com/openshift/enhancements/pull/1217#discussion_r982979140) as well to keep it map type.

Author: comment incorporated.

  ...
  labels = local.labels
}
```

For resources supporting labels in beta version below changes are required.
``` terraform
provider "googlebeta" {
  credentials = var.gcp_service_account
  project     = var.gcp_project_id
  region      = var.gcp_region
}

resource "google_compute_address" "cluster_ip" {
  provider = googlebeta
  ...
  labels = local.labels
}
```

The label of the form `kubernetes-io-cluster-<cluster_id>:owned` added by OpenShift, where
cluster_id is a string generated by concatenating user inputted cluster name and a random
string will be limited to a maximum length of 27 characters by trimming long cluster name
to 21 characters.

***GCP Tags*** 

Add a new field `resourceTags` to `.status.platformStatus.gcp` of the
`infrastructure.config.openshift.io` type. Tags included in the `resourceLabels` field
will be applied to new resources created for the cluster by the in-cluster operators.

The `resourceTags` field in `status.platformStatus.gcp` will be populated by the
installer using the entries from `platform.gcp.userTags` field of `install-config`.

`status.platformStatus.gcp.resourceTags` field of `infrastructure.config.openshift.io`
is immutable.

> OpenShift operators(Cluster Infrastructure, Storage) that create GCP resources will apply

Reviewer 3: Does Cluster API Provider GCP support applying tags?

Author: I think, it's not supported. I went through the reconciler code and couldn't find any references for user-defined tags, but they support defining network tags, which is present in machine-api as well.

Reviewer 3: It will need to be added upstream before we can accept it downstream in that case

Author: Changes in GCPMachineProviderSpec is required for users to configure additional tags apart from those provided during cluster creation in install-config. Since VMs will tagged with tags provided by user during installation and extending GCPMachineProviderSpec to support tags require upstream and due to the time limit, would it be acceptable to add this feature in GA and in TP it will be a limitation?

Author: Added as limitation for TP.

  3. google_compute_address API provides just the `labels` functionality in the beta version,
  which is in beta since the initial [commit](https://github.com/hashicorp/terraform-provider-google-beta/commit/0fdc262183ed82e78824e42d05042a7b0aba7c8b)
  in 2018, though the usual beta phase for a feature is six months.

### Drawbacks
- User-defined labels cannot be updated on an GCP resource which is not managed by an
  operator. In this proposal, the changes proposed and developed will be part of
  openshift-* namespace. External operators are not in scope.
  User-defined labels can be updated on the following GCP resources.
  - Compute Engine Instance
  - Compute Address
  - Compute Image
  - Compute Forwarding Rule
  - Storage Bucket
  - DNS Zone

- GCP Labels can be used as queryable annotations for resources, but can't be used
  to set conditions on policies.

- User-defined tags cannot be updated on a GCP resource which is not managed by an
  operator. In this proposal, the changes proposed and developed will be part of
  openshift-* namespace. External operators are not in scope.
  User-defined tags can be updated on the following GCP resources.
  - Compute Engine Instances

>   - Cloud Storage Buckets

Reviewer 2: If tags can only be created on two resources, this seems like a lot of effort to tag only these two. 

What is the motivation for adding tags in addition to labels? i.e. does enabling tags allow us to do something we can't do with just labels?

Reviewer 2: If we need to tag buckets created for persistent volumes, which should get review from @jsafrane

Author: Tags provide additional functionalities of able to set IAM policy conditions, Organization policy conditions, cost analysis and integration with cloud billing, whereas labels can be used for organizing resources and integration with cloud billing.

> If we need to tag buckets created for persistent volumes, which should get review from @jsafrane

Thank you for the suggestion! I will reach out to @jsafrane for the review.

  the operators will make use of rate limiting mechanism to limit the number requests
  generated per second.
  
- GCP client libraries support two ways of providing tag details
  1. NamespacedName: Tag value could be fed to APIs in the format
     `<ORGANIZATION_ID>/<TAG_KEY_SHORT_NAME>/<TAG_VALUE_SHORT_NAME>` where
      - ORGANIZATION_ID: Organization ID which is hosting the project under which OCP cluster
     will be created and tag key, value resources exist.
      - TAG_KEY_SHORT_NAME: Tag key short name is the display name provided during tag key
     resource creation.
      - TAG_VALUE_SHORT_NAME: Tag value short name is the display name provided during tag value
     resource creation.
  2. Name: Tag value could also be fed to the APIs in the format `tagValues/<TAG_VALUE_ID>` where
      - TAG_VALUE_ID: Tag value ID is the generated numeric ID for the tag value resource using which
     the tag key it is linked to can be identified.

  GCP terraform API `google_tags_location_tag_binding` expects the tag value's `Name` whereas the GO API
  `CreateTagBinding` accepts both tag value's `Name` and `NamespacedName`. The current approach for user

>   to input `ORGANIZATION_ID, TAG_KEY_SHORT_NAME, TAG_VALUE_SHORT_NAME` which has more readability, will require

Reviewer 2: Are we expecting the installer to create the tags? Or the users will pass in references to already existing tags?

Author: Users have to pass the references to the tags which already exist at the Organization where the project in which OpenShift cluster will be created is present.

}
```

- Add `OrganizationID`, `resourceLabels` and `resourceTags` fields to `status.platformStatus.gcp`
  of the `infrastructure.config.openshift.io`
```\h*go\K
// GCPPlatformStatus holds the current status of the Google Cloud Platform infrastructure provider.
// +kubebuilder:validation:XValidation:rule="!has(oldSelf.resourceLabels) && !has(self.resourceLabels) || has(oldSelf.resourceLabels) && has(self.resourceLabels)",message="resourceLabels may only be configured during installation"
// +kubebuilder:validation:XValidation:rule="!has(oldSelf.resourceTags) && !has(self.resourceTags) || has(oldSelf.resourceTags) && has(self.resourceTags)",message="resourceTags may only be configured during installation"
type GCPPlatformStatus struct {
	// organizationID is the Organization ID the new GCP resources created for the cluster.
	// belongs to. To find the Organization ID follow
	// https://cloud.google.com/resource-manager/docs/creating-managing-organization#retrieving_your_organization_id
	// +kubebuilder:validation:MinLength=1
	// +kubebuilder:validation:MaxLength=32
	// +kubebuilder:validation:Pattern=`^[1-9][0-9]+$`
	// +optional

> 	OrganizationID string `json:"organizationID"`

Reviewer 2: So this probably needs to be moved inside `GCPResourceTag` and perhaps changed from `OrganizationID` to `Parent` or `ParentID` so that we can attach tags from projects as well as organizations

Author: suggestion incorporated.

  `infrastructure.config.openshift.io` to add labels to the resources created later.

***GCP Tags***
- An OpenShift administrator requests to add required tags to all GCP resources
  created by OpenShift by adding it in `platform.gcp.userTags`.
- OpenShift installer validates the tags defined in `platform.gcp.userTags` and
  adds these tags to all resources created during installation and also updates
  `.status.platformStatus.gcp.resourceTags` of the `infrastructure.config.openshift.io`
- In-cluster operators refers `.status.platformStatus.gcp.resourceTags` of the
  `infrastructure.config.openshift.io` to add tags to the resources created later.

#### Variation [optional]

### API Extensions
Enhancement requires below modifications to the mentioned CRDs
- Add `userLabels` and `userTags` fields to `platform.gcp` of the
  `installconfigs.install.openshift.io`
```\h*go\K
type Platform struct {
	// userLabels has additional keys and values that the installer will add as
	// labels to all resources that it creates on GCP. Resources created by the
	// cluster itself may not include these labels. This is a TechPreview feature

> 	// and requires setting featureSet to TechPreviewNoUpgrade to configure labels.

Author: install-config has field [featureSet](https://github.com/openshift/installer/blob/master/data/data/install.openshift.io_installconfigs.yaml#L1670-L1682), which when set to `TechPreviewNoUpdate` creates featureGate with all the features marked as TechPreview. And this field is used for validating if userLabels and userTags in install-config could be defined. And for operators feature should be enabled using featureGates. Please let me know your thoughts.

	// GCP supports a maximum of 64 labels per resource. OpenShift reserves 32 labels for internal use,
	// allowing 32 labels for user configuration.
	// +kubebuilder:validation:MaxItems=32
	// +kubebuilder:validation:XValidation:rule="self.all(x, x in oldSelf) && oldSelf.all(x, x in self)",message="resourceLabels are immutable and may only be configured during installation"
	// +listType=map
	// +listMapKey=key
	// +optional
	ResourceLabels []GCPResourceLabel `json:"resourceLabels,omitempty"`

	// resourceTags is a list of additional tags to apply to GCP resources created for the cluster.
	// See https://cloud.google.com/resource-manager/docs/tags/tags-overview for information on
	// tagging GCP resources. GCP supports a maximum of 50 tags per resource.
	// +kubebuilder:validation:MaxItems=50
	// +kubebuilder:validation:XValidation:rule="self.all(x, x in oldSelf) && oldSelf.all(x, x in self)",message="resourceTags are immutable and may only be configured during installation"
	// +listType=map
	// +listMapKey=key
	// +optional
	ResourceTags []GCPResourceTag `json:"resourceTags,omitempty"`
}

// GCPResourceLabel is a label to apply to GCP resources created for the cluster.

> type GCPResourceLabel struct {

Reviewer 3: Should we be validating that these labels don't conflict with Kubernetes/OpenShift/HyperShift namespaces?

Author: HyperShift doesn't run on GCP and Kubernetes providers, as part of this feature adds labels filestore instances created by the upstream driver and takes care of giving precedence to it's own labels. And CAPI adds labels `capi-cluster-<name>` and `capi-role` and this too is handled by overwriting the conflicting labels. Current handling should suffice I think correct?

Reviewer 3: I would have expected the installer to reject any label that's in the kubernetes or openshift naming schemes since those are normally reserved, we did the same on AWS and Azure right?

	// key is the key part of the label. A label key can have a maximum of 63 characters and cannot be empty.
	// Label must begin with a lowercase letter, and must contain only lowercase letters, numeric characters,
	// and the following special characters `_-`.
	// +kubebuilder:validation:Required
	// +kubebuilder:validation:MinLength=1
	// +kubebuilder:validation:MaxLength=63
	// +kubebuilder:validation:Pattern=`^[a-z][0-9a-z_-]+$`
	Key string `json:"key"`

	// value is the value part of the label. A label value can have a maximum of 63 characters and cannot be empty.
	// Value must contain only lowercase letters, numeric characters, and the following special characters `_-`.
	// +kubebuilder:validation:Required
	// +kubebuilder:validation:MinLength=1
	// +kubebuilder:validation:MaxLength=63
	// +kubebuilder:validation:Pattern=`^[0-9a-z_-]+$`
	Value string `json:"value"`
}

// GCPResourceTag is a tag to apply to GCP resources created for the cluster.
type GCPResourceTag struct {

> 	// parentID is the ID of the hierarchical resource where the tags are defined
> 	// e.g. at the Organization or the Project level. To find the Organization or Project ID ref
> 	// https://cloud.google.com/resource-manager/docs/creating-managing-organization#retrieving_your_organization_id
> 	// https://cloud.google.com/resource-manager/docs/creating-managing-projects#identifying_projects

Reviewer 3: This needs a little more punctuation
```suggestion
	// parentID is the ID of the hierarchical resource where the tags are defined,
	// e.g. at the Organization or the Project level. To find the Organization or Project ID refer to the following pages:
	// https://cloud.google.com/resource-manager/docs/creating-managing-organization#retrieving_your_organization_id,
	// https://cloud.google.com/resource-manager/docs/creating-managing-projects#identifying_projects.
```

	// An OrganizationID must consist of decimal numbers, and cannot have leading zeroes.
	// A ProjectID must be 6 to 30 characters in length, can only contain lowercase letters, numbers,
	// and hyphens, and must start with a letter, and cannot end with a hyphen.
	// +kubebuilder:validation:Required
	// +kubebuilder:validation:MinLength=1
	// +kubebuilder:validation:MaxLength=32

> 	// +kubebuilder:validation:Pattern=`(^[1-9][0-9]{0,31}$)|(^[a-z][a-z0-9-]{4,28}[a-z0-9]$)`

Reviewer 3: Nice work on the regex here

be enabled to activate processing of user defined labels and tags.
```\h*go\K
func() {
	select {
	case <-featureGateAccessor.InitialFeatureGatesObserved():
		featureGates, err := featureGateAccessor.CurrentFeatureGates()
	case <-time.After(1 * time.Minute):
		log.Error(nil, "timed out waiting for FeatureGate detection")
		return nil, fmt.Errorf("timed out waiting for FeatureGate detection")
	}

	gcpLabelsTagsEnabled := featureGates.Enabled(configv1.FeatureGateGCPLabelsTags)

	if gcpLabelsTagsEnabled {
		// add user defined labels and tags
	}
}
```

#### Caveats
1. Updating or removing resource labels/tags added by OpenShift using an external interface,
   may or may not be reconciled by the operator managing the resource.
2. Updating labels/tags of individual resources is not supported and any label or tag present
   in `.status.platformStatus.gcp.resourceLabels` or `.status.platformStatus.gcp.resourceTags`
   respectively of `infrastructure.config.openshift.io/v1` resource will result in adding 
   labels and tags to all OpenShift managed GCP resources.
3. Limitations for TechPreview

>    - Worker compute machines created by `machine-api-provider-gcp` controller will not be tagged,
>      controller will be replaced with out-of-tree cloud provider for GCP, and requires tagging
>      support to be added in upstream.

Reviewer 3: This isn't quite right, and we don't have timelines for the replacement, so it needs to be added first upstream, then downstream
```suggestion
   - Worker compute machines created by `machine-api-provider-gcp` controller will not be tagged,
     controller will be replaced with cluster-api-provider-gcp, and requires tagging
     support to be added in upstream.
```
